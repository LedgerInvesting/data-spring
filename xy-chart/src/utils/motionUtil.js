export const scaleTo = (num, inMin, inMax, outMin, outMax) => {
  /**
  * Scales a value given the min/max scopes
  * @param {number} num - the value to be scaled
  * @param {number} inMin - min for the scale of num
  * @param {number} inMax - max for the scale of num
  * @param {number} outMin - outMin min for scale num is being to converted to
  * @param {number} outMax - outMax min for scale num is being to converted to
  * @return {number} - the scaled value of num
  */
  return (num - inMin) * (outMax - outMin) / (inMax - inMin) + outMin;
};

export const planeDistance = (x1, y1, x2, y2) => {
  /**
  * Calculate the distance between two sets of x/y coordinates
  * @param {number} x1 - x value for the first set of coordinates
  * @param {number} y1 - y value for the first set of coordinates
  * @param {number} x2 - x value for the second set of coordinates
  * @param {number} y2 - y value for the second set of coordinates
  * @return {number} - the distance between two points
  */
  const xDist = x1 - x2,
    yDist = y1 - y2;
  return Math.hypot(xDist, yDist);
};

export const originCalculator = (animationStart, point, bounds, scale, isArea, customAnimationStart) => {
  /**
  * Calculates the origin for an animated x/y value set
  * @param {string} animationStart - the type of origin to be calculated
  * @param {number} point.x - the x for the animating point
  * @param {number} point.y - the y for the animating point
  * @param {number} bounds.xMin - the x min bound
  * @param {number} bounds.yMin - the y min bound
  * @param {number} bounds.xMax - the x max bound
  * @param {number} bounds.yMax - the y max bound
  * @param {object} scale.x - a data-ui scale object, used in xy-chart range calculation
  * @param {object} scale.y - a data-ui scale object, used in xy-chart range calculation
  * @param {boolean} isArea - is the animating point part of an area series
  * @return {object} - The x and y values for the animationStart point
  */
  const originPoint = { ...point };
  switch (animationStart) {
    case 'bottom':
      if (isArea) {
        originPoint.y0 = bounds.yMin;
        originPoint.y1 = bounds.yMin;
      } else {
        originPoint.y = bounds.yMin;
      }
      return originPoint;
    case 'top':
      if (isArea) {
        originPoint.y0 = bounds.yMax;
        originPoint.y1 = bounds.yMax;
      } else {
        originPoint.y = bounds.yMax;
      }
      return originPoint;
    case 'left':
      return {
        ...originPoint,
        x: bounds.xMin,
      };
    case 'right':
      return {
        ...originPoint,
        x: bounds.xMax,
      };
    case 'bottomLeft':
      if (isArea) {
        originPoint.y0 = bounds.yMax;
        originPoint.y1 = bounds.yMax;
      } else {
        originPoint.y = bounds.yMax;
      }
      return {
        ...originPoint,
        x: bounds.xMin,
        y: bounds.yMin,
      };
    case 'bottomRight':
      if (isArea) {
        originPoint.y0 = bounds.yMin;
        originPoint.y1 = bounds.yMin;
      } else {
        originPoint.y = bounds.yMin;
      }
      return {
        ...originPoint,
        x: bounds.xMin,
      };
    case 'topRight':
      if (isArea) {
        originPoint.y0 = bounds.yMax;
        originPoint.y1 = bounds.yMax;
      } else {
        originPoint.y = bounds.yMax;
      }
      return {
        ...originPoint,
        x: bounds.xMax,
      };
    case 'topLeft':
      if (isArea) {
        originPoint.y0 = bounds.yMax;
        originPoint.y1 = bounds.yMax;
      } else {
        originPoint.y = bounds.yMax;
      }
      return {
        ...originPoint,
        x: bounds.xMin,
      };
    case 'edge':
      if (isArea) return point;
      const xRange = scale.x.range(),
        yRange = scale.y.range();

      const boundsPixels = {
        xMax: xRange[1],
        xMin: xRange[0],
        yMax: yRange[0],
        yMin: yRange[1],
      };

      const pointsPixels = {
        x: scaleTo(point.x, bounds.xMin, bounds.xMax, boundsPixels.xMin, boundsPixels.xMax),
        y: scaleTo(point.y, bounds.yMin, bounds.yMax, boundsPixels.yMin, boundsPixels.yMax),
      };

      // If the point is out of bounds, do not animate
      const outOfBounds = (pointsPixels.x > boundsPixels.xMax
        || pointsPixels.x < boundsPixels.xMin
        || pointsPixels.y > boundsPixels.yMax
        || pointsPixels.y < boundsPixels.yMin);
      if (outOfBounds) return point;

      // Edge points that are within a radial threshold of the corner will animate origin from the corner
      const cornerDistances = {
        bottomLeft: planeDistance(pointsPixels.x, pointsPixels.y, boundsPixels.xMin, boundsPixels.yMin),
        bottomRight: planeDistance(pointsPixels.x, pointsPixels.y, boundsPixels.xMax, boundsPixels.yMin),
        topRight: planeDistance(pointsPixels.x, pointsPixels.y, boundsPixels.xMax, boundsPixels.yMax),
        topLeft: planeDistance(pointsPixels.x, pointsPixels.y, boundsPixels.xMin, boundsPixels.yMax),
      };
      const cornerThreshold = (boundsPixels.xMax * 0.2);
      if (cornerDistances.bottomLeft < cornerThreshold) return { ...point, x: bounds.xMin, y: bounds.yMin };
      if (cornerDistances.bottomRight < cornerThreshold) return { ...point, x: bounds.xMax, y: bounds.yMin };
      if (cornerDistances.topRight < cornerThreshold) return { ...point, x: bounds.xMax, y: bounds.yMax };
      if (cornerDistances.topLeft < cornerThreshold) return { ...point, x: bounds.xMin, y: bounds.yMax };

      const edgeDist = {
        xMax: boundsPixels.xMax - pointsPixels.x,
        xMin: boundsPixels.xMin + pointsPixels.x,
        yMax: boundsPixels.yMax - pointsPixels.y,
        yMin: boundsPixels.yMin + pointsPixels.y,
      };
      const xEdge = edgeDist.xMax > edgeDist.xMin ? 'xMin' : 'xMax',
        yEdge = edgeDist.yMax > edgeDist.yMin ? 'yMin' : 'yMax',
        boundTo = edgeDist[xEdge] > edgeDist[yEdge] ? 'y' : 'x',
        closestEdge = (boundTo === 'y') ? yEdge : xEdge;

      return {
        ...point,
        [boundTo]: bounds[closestEdge],
      };
    case 'custom':
      const customCalculation = customAnimationStart(point);

      return {
        ...customCalculation,
      }
    default:
      return {
        ...point,
      };
  }
};

export const calcFlattenedStart = (props) => {
  /**
  * Calculates animation start values for an array of x/y objects
  * and returns a flattened dictionary of key/value pairs for react-spring animations
  * @param {array} - props.data - the data-ui series
  * @param {string} props.animationStart - the type of origin to be animated
  * @param {function} - props.sizeCalculator - function to be used to calculate a size start value for animation
  * @param props.bounds.xMin - the x min bound
  * @param props.bounds.yMin - the y min bound
  * @param props.bounds.xMax - the x max bound
  * @param props.bounds.yMax - the y max bound
  * @param {object} props.xScale - a data-ui scale object, used in xy-chart range calculation
  * @param {object} props.yScale - a data-ui scale object, used in xy-chart range calculation
  * @return {object} - an object representing the props.data series flattened with calculated animation origins
  */
  const {
    data, animationStart, sizeCalculator, bounds, xScale, yScale, customAnimationStart
  } = props;

  const motionPoints = {
    from: {},
    to: {},
  };
  if (data === undefined) return motionPoints;
  data.map((point, index) => {
    const isArea = point.hasOwnProperty('y0') && point.hasOwnProperty('y1');

    const origin = originCalculator(
      animationStart,
      point,
      bounds,
      { x: xScale, y: yScale },
      isArea,
      customAnimationStart,
    );
    motionPoints.from[`x${index}`] = origin.x;
    motionPoints.to[`x${index}`] = point.x;
    motionPoints.from[`y${index}`] = origin.y;
    motionPoints.to[`y${index}`] = point.y;

    if (isArea) {
      motionPoints.from[`y0${index}`] = origin.y0;
      motionPoints.to[`y0${index}`] = point.y0;
      motionPoints.from[`y1${index}`] = origin.y1;
      motionPoints.to[`y1${index}`] = point.y1;
    }
    if (sizeCalculator) {
      motionPoints.from.size = 1;
      motionPoints.to.size = sizeCalculator(point);
    }
    return point;
  });

  return motionPoints;
};

export const constructArray = (values, data, isArea) => {
  /** Constructs an array of x/y objects from a flattened object representation
  */
  return data.map((entry, index) => {
    const point = {
      ...entry,
      x: values[`x${index}`],
      y: values[`y${index}`],
    };
    if (isArea) {
      point.y0 = values[`y0${index}`];
      point.y1 = values[`y1${index}`];
    }
    return point;
  });
};
