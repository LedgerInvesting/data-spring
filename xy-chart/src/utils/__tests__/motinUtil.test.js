import React from 'react';
import { calcFlattenedStart, constructArray } from '../motionUtil';
import { config } from 'react-spring';
import renderer from 'react-test-renderer';


it('constructArray matches expected output for xySeries', () => {
  const testConstructArray = constructArray(
    flattenedValues.xySeries,
    testSeries.xySeries,
  );
  expect(testConstructArray).toMatchSnapshot();
});

it('constructArray matches expected output for boundArea series', () => {
  const testFlattenedStart = constructArray(
    flattenedValues.boundArea,
    testSeries.boundArea,
  );
  expect(testFlattenedStart).toMatchSnapshot();
});

it('calcFlattenedStart matches expected output for xySeries', () => {
  const testFlattenedStart = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'bottom',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  expect(testFlattenedStart).toMatchSnapshot();
});

it('calcFlattenedStart matches expected output for boundArea series', () => {
  const testFlattenedStart = calcFlattenedStart({
    data: testSeries.boundArea,
    animationStart: 'bottom',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  expect(testFlattenedStart).toMatchSnapshot();
});

it('calcFlattenedStart calculates all animationStart options', () => {
  const testBottomLeft = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'bottomLeft',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testBottomRight = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'bottomRight',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testTopRight = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'topRight',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testTopLeft = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'topLeft',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testTop = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'top',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testBottom = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'bottom',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testLeft = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'left',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testRight = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'right',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  const testEdge = calcFlattenedStart({
    data: testSeries.xySeries,
    animationStart: 'edge',
    sizeCalculator: () => 1,
    bounds: chartBounds,
    xScale: mockChartScale.xScale,
    yScale: mockChartScale.yScale,
  });
  expect(testBottomLeft).toMatchSnapshot();
  expect(testBottomRight).toMatchSnapshot();
  expect(testTopLeft).toMatchSnapshot();
  expect(testTopRight).toMatchSnapshot();
  expect(testBottom).toMatchSnapshot();
  expect(testTop).toMatchSnapshot();
  expect(testLeft).toMatchSnapshot();
  expect(testRight).toMatchSnapshot();
  expect(testEdge).toMatchSnapshot();
});

const testSeries = {
  boundArea: [
    {
      x: 2,
      y0: 0,
      y1: 6,
    },
    {
      x: 4,
      y0: 0,
      y1: 9,
    },
    {
      x: 5,
      y0: 0,
      y1: 6,
    },
    {
      x: 8,
      y0: 0,
      y1: 3,
    },
  ],
  xySeries: [
    {
      x: 5,
      y: 6,
    },
    {
      x: 8,
      y: 3,
    },
    {
      x: 2,
      y: 7,
    },
    {
      x: 10,
      y: 1,
    },
  ],
}

const flattenedValues = {
  xySeries: {
    y0: 2,
    x0: 1,
    y1: 4,
    x1: 5,
    y2: 6,
    x2: 8,
    y3: 9,
    x3: 15,
  },
  boundArea: {
    y00: 0,
    y10: 2,
    x0: 1,
    y01: 0,
    y11: 3,
    x1: 5,
    y02: 0,
    y12: 7,
    x2: 8,
    y03: 0,
    y13: 9,
    x3: 15,
  }
}

const chartBounds = {
  xMin: 0,
  yMin: 0,
  xMax: 9,
  yMax: 10,
};

const mockChartScale = {
  xScale: { range: () => [10, 0] },
  yScale: { range: () => [1, 10] }
}
