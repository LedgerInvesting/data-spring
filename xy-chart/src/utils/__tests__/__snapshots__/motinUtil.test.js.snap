// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calcFlattenedStart calculates all animationStart options 1`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 0,
    "x1": 0,
    "x2": 0,
    "x3": 0,
    "y0": 0,
    "y1": 0,
    "y2": 0,
    "y3": 0,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 2`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 9,
    "x1": 9,
    "x2": 9,
    "x3": 9,
    "y0": 0,
    "y1": 0,
    "y2": 0,
    "y3": 0,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 3`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 0,
    "x1": 0,
    "x2": 0,
    "x3": 0,
    "y0": 10,
    "y1": 10,
    "y2": 10,
    "y3": 10,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 4`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 9,
    "x1": 9,
    "x2": 9,
    "x3": 9,
    "y0": 10,
    "y1": 10,
    "y2": 10,
    "y3": 10,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 5`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 0,
    "y1": 0,
    "y2": 0,
    "y3": 0,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 6`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 10,
    "y1": 10,
    "y2": 10,
    "y3": 10,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 7`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 0,
    "x1": 0,
    "x2": 0,
    "x3": 0,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 8`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 9,
    "x1": 9,
    "x2": 9,
    "x3": 9,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart calculates all animationStart options 9`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`calcFlattenedStart matches expected output for boundArea series 1`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 2,
    "x1": 4,
    "x2": 5,
    "x3": 8,
    "y00": 0,
    "y01": 0,
    "y02": 0,
    "y03": 0,
    "y10": 0,
    "y11": 0,
    "y12": 0,
    "y13": 0,
  },
  "to": Object {
    "size": 1,
    "x0": 2,
    "x1": 4,
    "x2": 5,
    "x3": 8,
    "y00": 0,
    "y01": 0,
    "y02": 0,
    "y03": 0,
    "y10": 6,
    "y11": 9,
    "y12": 6,
    "y13": 3,
  },
}
`;

exports[`calcFlattenedStart matches expected output for xySeries 1`] = `
Object {
  "from": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 0,
    "y1": 0,
    "y2": 0,
    "y3": 0,
  },
  "to": Object {
    "size": 1,
    "x0": 5,
    "x1": 8,
    "x2": 2,
    "x3": 10,
    "y0": 6,
    "y1": 3,
    "y2": 7,
    "y3": 1,
  },
}
`;

exports[`constructArray matches expected output for boundArea series 1`] = `
Array [
  Object {
    "x": 1,
    "y0": 0,
    "y1": 2,
  },
  Object {
    "x": 5,
    "y0": 0,
    "y1": 3,
  },
  Object {
    "x": 8,
    "y0": 0,
    "y1": 7,
  },
  Object {
    "x": 15,
    "y0": 0,
    "y1": 9,
  },
]
`;

exports[`constructArray matches expected output for xySeries 1`] = `
Array [
  Object {
    "x": 1,
    "y": 2,
  },
  Object {
    "x": 5,
    "y": 4,
  },
  Object {
    "x": 8,
    "y": 6,
  },
  Object {
    "x": 15,
    "y": 9,
  },
]
`;
